download/src/lib.rs

```diff
@@ -35,108 +35,87 @@ pub enum Event<'a> {
     DownloadDataReceived(&'a [u8]),
 }
 
+pub type Callback<'a> = &'a dyn Fn(Event<'_>) -> anyhow::Result<()>;
+
 fn download_with_backend(
     backend: Backend,
     url: &Url,
     resume_from: u64,
-    callback: &dyn Fn(Event<'_>) -> Result<()>,
+    callback: Callback<'_>,
 ) -> Result<()> {
     match backend {
         Backend::Curl => curl::download(url, resume_from, callback),
         Backend::Reqwest(tls) => reqwest_be::download(url, resume_from, callback, tls),
     }
 }
 
-type DownloadCallback<'a> = &'a dyn Fn(Event<'_>) -> Result<()>;
-
 pub fn download_to_path_with_backend(
     backend: Backend,
     url: &Url,
     path: &Path,
     resume_from_partial: bool,
-    callback: Option<DownloadCallback<'_>>,
+    callback: Option<Callback<'_>>,
 ) -> Result<()> {
     use std::cell::RefCell;
     use std::fs::remove_file;
     use std::fs::OpenOptions;
     use std::io::{Read, Seek, SeekFrom, Write};
 
-    || -> Result<()> {
-        let (file, resume_from) = if resume_from_partial {
+    let (file, resume_from) = {
+        let mut downloaded_so_far = 0;
+
+        if resume_from_partial {
             let possible_partial = OpenOptions::new().read(true).open(path);
 
-            let downloaded_so_far = if let Ok(mut partial) = possible_partial {
+            if let Ok(mut partial) = possible_partial {
                 if let Some(cb) = callback {
-                    cb(Event::ResumingPartialDownload)?;
+                    cb(Event::ResumingPartialDownload).map_err(DownloadError::CallbackError)?;
 
                     let mut buf = vec![0; 32768];
-                    let mut downloaded_so_far = 0;
                     loop {
                         let n = partial.read(&mut buf)?;
-                        downloaded_so_far += n as u64;
                         if n == 0 {
                             break;
                         }
-                        cb(Event::DownloadDataReceived(&buf[..n]))?;
+                        downloaded_so_far += n;
+                        cb(Event::DownloadDataReceived(&buf[..n]))
+                            .map_err(DownloadError::CallbackError)?;
                     }
-
-                    downloaded_so_far
                 } else {
                     let file_info = partial.metadata()?;
-                    file_info.len()
-                }
-            } else {
-                0
-            };
+                    downloaded_so_far = file_info.len() as usize;
+                };
+            }
+        }
 
-            let mut possible_partial = OpenOptions::new()
-                .write(true)
-                .create(true)
-                .open(path)
-                .context("error opening file for download")?;
+        let mut possible_partial = OpenOptions::new().write(true).create(true).open(path)?;
```

Comment:
This drops the .context - will the resulting error be shown to users? The problem with file open errors is that they can come from anywhere...